sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,585 @@
+     * configuration.
+     */
+    public Read withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      checkNotNull(configuration, "MqttConnectionConfiguration");
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the max number of records received by the {@link Read}.
+     * When this max number of records is lower then {@code Long.MAX_VALUE}, the {@link Read}
+     * will provide a bounded {@link PCollection}.
+     *
+     * @param maxNumRecords The max number of records received.
+     * @return The {@link Read} {@link PTransform} with the corresponding max num records
+     * configuration.
+     */
+    public Read withMaxNumRecords(long maxNumRecords) {
+      checkArgument(maxReadTime() == null,
+          "maxNumRecord and maxReadTime are exclusive");
+      return toBuilder().setMaxNumRecords(maxNumRecords).build();
+    }
+
+    /**
+     * Define the max read time (duration) while the {@link Read} will receive messages.
+     * When this max read time is not null, the {@link Read} will provide a bounded
+     * {@link PCollection}.
+     *
+     * @param maxReadTime The max read time duration.
+     * @return The {@link Read} {@link PTransform} with the corresponding max read time
+     * configuration.
+     */
+    public Read withMaxReadTime(Duration maxReadTime) {
+      checkArgument(maxNumRecords() == Long.MAX_VALUE,
+          "maxNumRecord and maxReadTime are exclusive");
+      return toBuilder().setMaxReadTime(maxReadTime).build();
+    }
+
+    @Override
+    public PCollection<byte[]> apply(PBegin input) {
+
+      org.apache.beam.sdk.io.Read.Unbounded<byte[]> unbounded =
+          org.apache.beam.sdk.io.Read.from(new UnboundedMqttSource(this));
+
+      PTransform<PBegin, PCollection<byte[]>> transform = unbounded;
+
+      if (maxNumRecords() != Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords());
+      } else if (maxReadTime() != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime());
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      // validation is performed in the MqttConnectionConfiguration create()
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      mqttConnectionConfiguration().populateDisplayData(builder);
+      builder.add(DisplayData.item("maxNumRecords", maxNumRecords()));
+      builder.addIfNotNull(DisplayData.item("maxReadTime", maxReadTime()));
+    }
+
+  }
+
+  private static class UnboundedMqttSource
+      extends UnboundedSource<byte[], UnboundedSource.CheckpointMark> {
+
+    private final Read spec;
+
+    public UnboundedMqttSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
```

Comment:
Please fix all usages of raw types in the PR - there's several, I might have missed some.